Mathlib/LinearAlgebra/FreeModule/Finite/Matrix.lean

```diff
@@ -21,77 +21,93 @@ We provide some instances for finite and free modules involving matrices.
-    Module.Free R (M →ₗ[R] N) := by
-  cases subsingleton_or_nontrivial R
-  · apply Module.Free.of_subsingleton'
-  classical exact
-      Module.Free.of_equiv (LinearMap.toMatrix (chooseBasis R M) (chooseBasis R N)).symm
+instance Module.Free.linearMap [Module.Free S N] : Module.Free S (M →ₗ[R] N) :=
+  Module.Free.of_equiv (linearMapEquivFun R S M N).symm
 #align module.free.linear_map Module.Free.linearMap
 
-variable {R}
-
-instance Module.Finite.linearMap [Module.Finite R M] [Module.Finite R N] :
-    Module.Finite R (M →ₗ[R] N) := by
-  cases subsingleton_or_nontrivial R
-  · infer_instance
-  classical
-    have f := (LinearMap.toMatrix (chooseBasis R M) (chooseBasis R N)).symm
-    exact Module.Finite.of_surjective f.toLinearMap (LinearEquiv.surjective f)
+instance Module.Finite.linearMap [Module.Finite S N] : Module.Finite S (M →ₗ[R] N) :=
+  Module.Finite.equiv (linearMapEquivFun R S M N).symm
 #align module.finite.linear_map Module.Finite.linearMap
 
-end CommRing
+variable [StrongRankCondition R] [StrongRankCondition S] [Module.Free S N]
 
-section Integer
+open Cardinal
+theorem FiniteDimensional.rank_linearMap :
+    Module.rank S (M →ₗ[R] N) = lift.{w} (Module.rank R M) * lift.{v} (Module.rank S N) := by
+  rw [(linearMapEquivFun R S M N).rank_eq, rank_fun_eq_lift_mul,
+    ← finrank_eq_card_chooseBasisIndex, ← finrank_eq_rank R, lift_natCast]
 
-variable [AddCommGroup M] [Module.Finite ℤ M] [Module.Free ℤ M]
+/-- The finrank of `M →ₗ[R] N` as an `S`-module is `(finrank R M) * (finrank S N)`. -/
+theorem FiniteDimensional.finrank_linearMap :
+    finrank S (M →ₗ[R] N) = finrank R M * finrank S N := by
+  simp_rw [finrank, rank_linearMap, toNat_mul, toNat_lift]
+#align finite_dimensional.finrank_linear_map FiniteDimensional.finrank_linearMap
 
-variable [AddCommGroup N] [Module.Finite ℤ N] [Module.Free ℤ N]
+variable [Module R S] [SMulCommClass R S S]
 
-instance Module.Finite.addMonoidHom : Module.Finite ℤ (M →+ N) :=
-  Module.Finite.equiv (addMonoidHomLequivInt ℤ).symm
-#align module.finite.add_monoid_hom Module.Finite.addMonoidHom
+theorem FiniteDimensional.rank_linearMap_self :
+    Module.rank S (M →ₗ[R] S) = lift.{u'} (Module.rank R M) := by
+  rw [rank_linearMap, rank_self, lift_one, mul_one]
 
-instance Module.Free.addMonoidHom : Module.Free ℤ (M →+ N) :=
-  letI : Module.Free ℤ (M →ₗ[ℤ] N) := Module.Free.linearMap _ _ _
-  Module.Free.of_equiv (addMonoidHomLequivInt ℤ).symm
-#align module.free.add_monoid_hom Module.Free.addMonoidHom
+theorem FiniteDimensional.finrank_linearMap_self : finrank S (M →ₗ[R] S) = finrank R M := by
+  rw [finrank_linearMap, finrank_self, mul_one]
 
-end Integer
+end Ring
 
-section CommRing
+section AlgHom
 
-variable [CommRing R] [StrongRankCondition R]
+variable (K M : Type*) (L : Type v) [CommRing K] [Ring M] [Algebra K M] [Module.Free K M]
+  [Module.Finite K M] [CommRing L] [IsDomain L] [StrongRankCondition L] [Algebra K L]
```

Comment:
By the way, here and in other files there are places where we assume both CommRing and StrongRankCondition because the file doesn't import [commRing_strongRankCondition](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/StrongRankCondition.html#commRing_strongRankCondition) (maybe they predate the formalization of the theorem). Do you think it's a good idea to import Mathlib.LinearAlgebra.FreeModule.StrongRankCondition and remove these redundant StrongRankConditions?

Reply:
IIRC I made this choice exactly to keep the imports small, but now the import structure changed a lot so maybe it is less reasonable. Anyway a comment is surely a good idea.

Reply:
I see, added comment.

Reply:
Actually it was maybe because it created a loop, so we had no choice. Does it work now?

Reply:
I didn't check this file but I [introduced](https://github.com/leanprover-community/mathlib4/commit/a3e8d01e04f285df56c4b2a60eaf393b3fceb2ac#diff-1a470a10fd05f0eedb570bfbbe37c36b1a2472e9dc4c78aa7fc070d33ec2c2d2) the import into LinearAlgebra/Dual and it works.

Reply:
If it works here I think we can go ahead.

Reply:
Let me just try it ...